### PR TITLE
Remove "monaco-editor/react", "react-monaco-editor" from the dependency list

### PR DESCRIPTION
**Pull request by Henry8192**
# References
Junhao found unneeded dependencies for Monaco Editor. Further investigation led to believe that non-official dependencies [@monaco-editor/react](https://github.com/suren-atoyan/monaco-react) and [react-monaco-editor](https://github.com/react-monaco-editor/react-monaco-editor) can be removed if we implement our own wrapper for Microsoft's official [monaco-editor](https://github.com/microsoft/monaco-editor), which will enable us to use the latest release from Microsoft. 

# Description
## 1. Removal of [react-monaco-editor](https://github.com/react-monaco-editor/react-monaco-editor)
Current codebase has no reference of the package. 

## 2. Removal of [monaco-editor/react](https://github.com/suren-atoyan/monaco-react)
In `src/Viewer/components/Monaco/MonacoInstance.js`:
- `handleEditorWillMount`, `handleEditorDidMount`,  `beforeMount` and `onMount` do not need any parameters, so the parameters are removed.
- Instead of returning an Editor element with passed-in arguments and letting dependency to handle the configuration, we directly create a Monaco Editor instance within the code, apply the configuration and return it.

# Validation performed
1. Remove `node_modules` and `package-lock.json`, and do `npm install`.
2. `npm start` and open a zst file.
3. The browser successfully shows the log. Performed scrolling and switching themes. No errors observed in the browser's inspect element page.
4. Load another IR Stream file to see if on-the-fly editor value setting is working.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -19,7 +19,7 @@
   },
   "homepage": "https://github.com/y-scope/yscope-log-viewer#readme",
   "dependencies": {
-    "@monaco-editor/react": "^4.4.6",
+    "monaco-editor-webpack-plugin": "^7.1.0",
```
Comment (junhaoliao) [inline package.json:22]: Sort the properties. e.g., In WebStorm, you can 
image

File: src/Viewer/components/Monaco/MonacoInstance.js [hunk 2]
```diff
@@ -1,15 +1,12 @@
-import React, {useContext, useEffect, useRef, useState} from "react";
-
-import Editor, {loader} from "@monaco-editor/react";
+import React, { useContext, useEffect, useRef, useState } from "react";
 import * as monaco from "monaco-editor";
 import PropTypes from "prop-types";
 
-import {THEME_STATES} from "../../../ThemeContext/THEME_STATES";
-import {ThemeContext} from "../../../ThemeContext/ThemeContext";
+import { THEME_STATES } from "../../../ThemeContext/THEME_STATES";
```
Comment (junhaoliao) [inline src/Viewer/components/Monaco/MonacoInstance.js:5]: Let's apply the eslint config file so that when we use a code formatter, we can avoid changing lines that are unrelated to the PR.

File: src/Viewer/components/Monaco/MonacoInstance.js [hunk 3]
```diff
@@ -250,9 +236,8 @@ function MonacoInstance ({
     }, [logData]);
 
     useEffect(() => {
-        setMonacoTheme((theme === THEME_STATES.LIGHT)
-            ?"customLogLanguageLight"
-            :"customLogLanguageDark");
+        const monacoTheme = (theme === THEME_STATES.LIGHT) ? "customLogLanguageLight" : "customLogLanguageDark"
```
Comment (junhaoliao) [inline src/Viewer/components/Monaco/MonacoInstance.js:239]: Let's make this a named helper function. i.e.,
```
const getMonacoThemeName = (theme) => (
    (theme === THEME_STATES.LIGHT) ? "customLogLanguageLight" : "customLogLanguageDark"
);
```

File: src/Viewer/components/Monaco/MonacoInstance.js [hunk 4]
```diff
@@ -272,23 +257,22 @@ function MonacoInstance ({
...
-        "readOnly": true,
-        "renderWhitespace": "none",
-        "wordWrap": "on",
-        "scrollBeyondLastLine": false,
-    };
+    useEffect(() => {
+        if (null === editorRef.current) {
+            handleEditorWillMount();
+            editorRef.current = monaco.editor.create(editorContainerRef.current, {
+                language: "logLanguage",
```
Comment (junhaoliao) [inline src/Viewer/components/Monaco/MonacoInstance.js:264]: Sort the properties

File: src/Viewer/components/Monaco/MonacoInstance.js [hunk 5]
```diff
@@ -272,23 +257,22 @@ function MonacoInstance ({
...
-        <Editor
-            defaultValue="Loading content..."
-            theme={monacoTheme}
-            language={language}
-            beforeMount={handleEditorWillMount}
-            onMount={handleEditorDidMount}
-            options={monacoOptions}
-        />
-    );
+    return (<div ref={editorContainerRef} style={{ display: 'flex', width: '100% ', height: "100%" }}></div>)
```
Comment (junhaoliao) [inline src/Viewer/components/Monaco/MonacoInstance.js:275]: Question: Do we have to specify the width and height to be 100% of the parent container size? Just curious if the monaco.editor.create() function will set the styles on the DOM.

**Review by junhaoliao — CHANGES_REQUESTED** [review]
Thanks for making the changes to effectively reduce our package size and enables us to use the latest release from Microsoft. 

Please help address the comments posted. Thanks!

File: src/Viewer/components/Monaco/MonacoInstance.js (hunk 5 above)
Comment (Henry8192) [reply src/Viewer/components/Monaco/MonacoInstance.js:275]: Without the manually configure the style, the editor is not shown properly :(
See the attachments after I commented out the width and height
Screenshot_20240119_171009

Comment (junhaoliao) [reply src/Viewer/components/Monaco/MonacoInstance.js]: Gotcha. Let's keep the styles then. Please consider putting the styles for a class into a .css file and set the classname on the component.

Comment (Henry8192) [reply src/Viewer/components/Monaco/MonacoInstance.js]: Should we create a new .css file or use the MonacoInstance.scss?

Comment (junhaoliao) [reply src/Viewer/components/Monaco/MonacoInstance.js]: MonacoInstance.scss please.
We can name it like `monaco-container`

File: src/Viewer/components/Monaco/MonacoInstance.js [hunk 6]
```diff
@@ -272,22 +257,23 @@ function MonacoInstance ({
...
     }, []);
 
-    const monacoOptions = {
-        "readOnly": true,
-        "renderWhitespace": "none",
-        "wordWrap": "on",
-        "scrollBeyondLastLine": false,
-    };
+    useEffect(() => {
+        if (null === editorRef.current) {
```
Comment (junhaoliao) [inline src/Viewer/components/Monaco/MonacoInstance.js:261]: Change L222: 
```
if (editorRef && editorRef.current)
```
->
```
if (null != editorRef.current)
```

editorRef is always an object and therefore always not null.

File: src/Viewer/components/Monaco/MonacoInstance.js [hunk 7]
```diff
@@ -100,21 +98,17 @@ function MonacoInstance ({
...
     const {theme} = useContext(ThemeContext);
     const editorRef = useRef(null);
-    const monacoRef = useRef(null);
     const timeoutRef = useRef(null);
-    const [monacoTheme, setMonacoTheme] = useState("customLogLanguageLight");
-
-    const [language, setLanguage] = useState("");
-
-    loader.config({monaco});
+    const editorContainerRef = useRef(null);
```
Comment (junhaoliao) [inline src/Viewer/components/Monaco/MonacoInstance.js:102]: Might make slightly more sense to move this editorContainerRef declaration directly below editorRef

File: src/Viewer/components/Monaco/MonacoInstance.js [hunk 8]
```diff
@@ -100,21 +98,17 @@ function MonacoInstance ({
...
 
     /**
      * Called before the monaco editor is mounted.
-     *
-     * @param {object} monaco
      */
-    function handleEditorWillMount (monaco) {
-        beforeMount(monaco);
+    const handleEditorWillMount = () => {
+        beforeMount();
```
Comment (junhaoliao) [inline src/Viewer/components/Monaco/MonacoInstance.js:111]: Now we have identified that argument `monaco` is not needed, let's also update type `BeforeMountCallback` in the docstring in L67.

File: src/Viewer/components/Monaco/MonacoInstance.js [hunk 9]
```diff
@@ -170,12 +158,11 @@ function MonacoInstance ({
                 }, 50);
             }
         });
-
-        onMount(editor, monaco);
+        onMount();
```
Comment (junhaoliao) [inline src/Viewer/components/Monaco/MonacoInstance.js:161]: ditto, let's also update type MountCallback.

File: src/Viewer/components/Monaco/MonacoInstance.js [hunk 10]
```diff
@@ -137,22 +131,16 @@ function MonacoInstance ({
...
     /**
      * Called when editor is finished mounting.
-     *
-     * @param {object} editor
-     * @param {object} monaco
      */
-    const handleEditorDidMount =(editor, monaco) => {
-        monacoRef.current = monaco;
-        editorRef.current = editor;
+    const handleEditorDidMount = () => {
```
Comment (junhaoliao) [inline src/Viewer/components/Monaco/MonacoInstance.js:140]: Consider renaming this to `_onMount` so we are consistent with the name in the component props.

File: src/Viewer/components/Monaco/MonacoInstance.js [hunk 11]
```diff
@@ -100,21 +98,17 @@ function MonacoInstance ({
...
+    );
 
     /**
      * Called before the monaco editor is mounted.
-     *
-     * @param {object} monaco
      */
-    function handleEditorWillMount (monaco) {
-        beforeMount(monaco);
+    const handleEditorWillMount = () => {
```
Comment (junhaoliao) [inline src/Viewer/components/Monaco/MonacoInstance.js:110]: Consider renaming this to `_beforeMount` so we are consistent with the name in the component props.

File: package.json [hunk 12]
```diff
@@ -49,7 +47,7 @@
     "@babel/preset-react": "^7.18.6",
     "babel-loader": "^9.1.0",
     "css-loader": "^6.7.1",
-    "eslint": "8.22.0",
+    "eslint": "^8.22.0",
```
Comment (junhaoliao) [inline package.json:50]: I believe that caret symbol was not added under an assumption that the linter behaviour might change in a newer version. We can refer to this section to understand the use of the caret (`^`) symbol: https://github.com/npm/node-semver?tab=readme-ov-file#caret-ranges-123-025-004

**Review by junhaoliao — CHANGES_REQUESTED** [review]
More improvements identified. Please help address. Thanks!

File: src/Viewer/components/Monaco/MonacoInstance.js (hunk 6 above)
Comment (Henry8192) [reply src/Viewer/components/Monaco/MonacoInstance.js:261]: != or !== ?

File: package.json (hunk 12 above)
Comment (Henry8192) [reply package.json:50]: Sure. It is rare that ESLint extension in VS Code automatically changes this for me.

File: src/Viewer/components/Monaco/MonacoInstance.js (hunk 6 above)
Comment (junhaoliao) [reply src/Viewer/components/Monaco/MonacoInstance.js:258]: My aplologies - should be `!==`. Thanks for correcting another occurance!

**Comment by junhaoliao** [comment]
Thanks. Looks good to me. Adding Kirk for final reviews.
